### ISSUE TITLE
in depth Exult combat analysis

### ISSUE THREAD

**Issue by DominusExult** [exult/exult#178]
@KarlClinckspoor made a very insightful in depth combat analysis of Exult - see https://karlclinckspoor.github.io/2022/04/10/exult_combat.html

A great read! Thank you!

**Comment by drcode1** [comment]
I skimmed it just now, and am amazed at the detail.  I notice that at 
the top, he mentions that Exult's combat is simplified vs the original.  
We've all thought about how we can improve that, but it's always tricky 
to make changes.

**Comment by DominusExult** [comment]
To impr Exult's combat I'd say the first step would be to implement the party member strategies (protect, flank, flee, rush...) by looking closely at how the original does that.
And then see what to reuse for attackers.
Of course, the original's combat looked like a mess, but it actually had great ideas and once you pay attention it was well thought.

For example your code, I had to comment unfortunately for #147, was already making it look much nicer. As it gave the hint of the party following a strategy ;)

**Comment by drcode1** [comment]
I'd like to re-visit that change I did.  Seems like it might work to 
have 'approach_foe' tried first, and only if it fails several times, do 
'wander_for_attack'.  It's good that I have the Fawn Tower example to 
try it on.

**Comment by Vidsek** [comment]
My main issue with current combat AI is how *all* the party clumps right on the mob(s) being attacked.  Party members (and mob NPCs) equipped with ranged weapons should stand back a ways (but within their weapon's range) and ones equipped with melee weapons should stand at the max range of their weapon (mostly significant for halbards and, especially, whips).
This would hopefully reduce the incidence of friendly fire injuries (mainly from fire type attacks that produce a fire field on the terrain) and, of course, *look* better X )

**Comment by DominusExult** [comment]
I spent some time the last few days with combat and wander_for_attack (#147).
I still haven't figured out why it fails at times but it brought me to some ideas how to make combat a bit more enjoyable. Especially it made me realize that we *actually do* support most of the combat modes you can set party members to.

- wander_for_attack could be made into the attack mode *flank* and also be used for ranged weapons. This code in combat.cc:`Combat_schedule::start_strike` works (with the problem that if *all* party members are set to flank/use range weapons, the combat breaks again as in #147) :
```
if (npc->get_attack_mode() == Actor::flank || winf->get_uses() == Weapon_info::ranged)
   wander_for_attack();
else
   approach_foe(true); // Try to get adjacent.
```
- The default combat mode is `Actor::nearest` which makes our combat look so terrible as the whole party gangs up on one enemy all the time. I propose that we need to fine tune combat.cc `Game_object *Combat_schedule::find_foe`.
Once an opponent is found you should check whether that one is already the target of another party member and if so search again for an opponent. Of course this should be limited to three or even just two tries.
IMO this would nee to be done for case Actor::nearest/Actor::random:/default:
Coding this is out of my abilities, though, but would benefit the look of our combat immensely :)
 @drcode1 are you up for that?

**Comment by drcode1** [comment]
On 2/24/2023 4:02 AM, Dominik Reichardt wrote:
>
>   * IMO this would nee to be done for case
>     Actor::nearest/Actor::random:/default:
>     Coding this is out of my abilities, though, but would benefit the
>     look of our combat immensely :)
>     @drcode1 <https://github.com/drcode1> are you up for that?
>
Yes, I've been wanting to do some work on Exult again.  Will be away for 
a few weeks with only my work computer, but should have time when I get 
back.

**Comment by drcode1** [comment]
I spent some time with the 'wander' code today, and did my first commit 
in what seems like years. :-)  My goal is to get the benefit of 
'wander_for_attack' without the problems that were reported.  I've only 
tested on Fawn Tower, so let me know of any problems, and feel free to 
fiddle with it.

**Comment by DominusExult** [comment]
Thank you @drcode1 coupled with the better selcting of targets for nearest and random combat looks way better. wander_for_attack still stalls sometimes, though, but @marzojr has some ideas to make the whole function work better.

**Comment by marzojr** [comment]
I added some improvements to target selection that might help make the combat modes feel different form each other, in case anyone wants to try.

Regarding the changes from wander_for_attack, I am going to do a write up of my current thoughts here to get some feedback.

The first thing you can observe from the original is that NPCs attacking in melee always tend to attack from maximum range for the weapon, repeatedly backing off after each strike. This makes pikemen very annoying to fight without ranged weapons or pikes/halberds of your own.

So for ranged combat: first part of the "flow chart" would be:

1. If the NPC has a target within range and a line of sight, attack
2. If the NPC has a target within range but no line of sight:
    a. If the object blocking line of sight is also an enemy, swap targets to it and attack
    b. Otherwise, proceed to find a line of sight
3. If the NPC has an out-of-range target, proceed to get in-range and get a line of sight
4. Otherwise, find a target and start over

Logic for finding an in-range spot with line of sight:

1. Find desired range. This would be a balance between the max range of the weapon, and the chance to hit the opponent (given the ranged attack penalties).
2. Create a ring of tiles centered at the target with radius equal to the desired range. Maybe we want to create a half-ring instead?
3. Start processing the tiles of this ring in order of proximity (using simple tile distance) to the NPC (so there would be less walking before an attack).
4. If the tile is blocked, skip it (dirt cheap check).
6. If there is no line of sight to the opponent (relatively cheap because it is a straight line check):
    a. If the object blocking line of sight is also an enemy, swap targets to it and attack
    b. Otherwise, skip the tile
5. If there is an enemy within (twice?) melee weapon range of the tile, skip it (involves distance checks and weapon data lookups, so it can cost a bit).
7. if the tile is not reachable, skip it (the more expensive check).
8. otherwise, use the path found to go to the target tile.
9. if there are no more tiles left in the ring, try a ring with a smaller radius on the next tick.

Another alternative would be to invert the check: find lines of sight from the **opponent** out to tiles of the ring, in the same processing order as before. This could be done in a way that gives a "bitmap" of tiles where the opponent can be attacked from, so that even if no tiles from the ring are available, we still have a set of tiles with line of sight which could be processed by the above algorithm.

Also, regarding flanking attack mode: currently (as of today!), we go for opponents that are also being attacked by an ally. I am thinking that maybe we also should pick a position behind the opponent, or maybe to the side, when approaching to attack them. From the sides would work wonders in ranged mode because the NPC would need to walk less; for melee, going for the back would be very useful against halberdiers. Or maybe there could be even more coordination, with successive NPCs picking different attack vectors on the same opponent.

Any thoughts?

**Comment by DominusExult** [comment]
I think somewhere in this sound logic we would need to factor in being attacked by close range enemy while the target is far off. BUT if we factor this in the chances are that the much nicer looking combat is falling back to a cluster fight which looks annoying.

**Comment by marzojr** [comment]
Another thing I mentioned in Discord which is good to have here: maybe with this logic we might want to try and see if friendly fire can be activated for ranged weapons, as NPCs will try to avoid hitting friends anyway. If it is still bad, we can reserve it for the higher combat difficulties 😈

On the converse, maybe we want to disable friendly fire on explosions to match the way projectiles work (as well as making weapons like the firedoom staff not be suicidal weapons).

**Comment by DominusExult** [comment]
Yes, that! And of course enable friendly explosion hits in Hard Combat.

Another idea for hard combat would be enabling the original fleeing while losing inventory items (and do we do the bloody footprints/blood pools when fleeing?).

**Comment by marzojr** [comment]
We don't do bloody footprints, no. Looking at the cheat screen in the original, there is a 'bleeding' flag which we do not implement which is probably responsible for this.

After trying out a bit, this flag is set whenever you take enough damage to drop a blood pool; but setting it (manually or by taking damage), it does not seem to do anything when moving around. After experimenting a bit, it seems to only cause bleeding when you are fleeing from taking a lot of damage (flee strategy does not bleed even when you are actually fleeing). We might want to change all of this in Exult, and just bleed while the flag is set.

By pure luck I discovered that 'bleeding' flag is cleared when entering combat (mainly, because I had manually set it on all party members, entered combat, and it was cleared...), but it does not seem to be cleared by waiting around (I left the game running for a few in-game days and it was still set) or by healing. Again we might want to change all that in Exult; maybe by having a timer for it, **and** maybe clear it also when the NPC is healed.
